### PR TITLE
(#44) feat: profile added

### PR DESCRIPTION
**Pull request by seunggabi**
### context
- #44 

### test 
```shell
cat ~/.dhapi/credentials    
[default]
username = seunggabi
password = ***********

[test]
username = asdf
password = 1234
```
```shell
PYTHONPATH=./src/ python3 src/dhapi/main.py buy_lotto645 -q  

[Lotto645 Buy Request]
Game A: []
Game B: []
Game C: []
Game D: []
Game E: []

❓ 위와 같이 구매하시겠습니까? [Y/n] yes
✅ --quiet 플래그가 주어져 자동으로 구매를 진행합니다.
2023-05-31 21:47:34 [error    ] 🚨 에러가 발생했습니다: 구매에 실패했습니다: [온라인복권 주간 구매한도] 초과되었습니다.
🚨 에러가 발생했습니다: 구매에 실패했습니다: [온라인복권 주간 구매한도] 초과되었습니다.

PYTHONPATH=./src/ python3 src/dhapi/main.py buy_lotto645 -q -p qwer
2023-05-31 21:47:38 [error    ] 🚨 에러가 발생했습니다: 로그인에 실패했습니다. 아이디 또는 비밀번호를 확인해주세요.
🚨 에러가 발생했습니다: 로그인에 실패했습니다. 아이디 또는 비밀번호를 확인해주세요.
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # DongHaeng Lottery API (Unofficial)
 
-```sh
+```shell
```
Comment (roeniss) [inline README.md:3]: Just curious -- why do you change this? Is this a more formal way? If so, Do you have documentation/evidence on it?

File: src/dhapi/router/arg_parser.py [hunk 2]
```diff
@@ -72,21 +73,23 @@ def __init__(self):
...
+            default=["default"],
+            help="https://github.com/roeniss/dhlottery-api#계정 정보 세팅",
         )
-        buy_lotto645.add_argument("-d", "--debug", action="store_true", help="로그 출력 레벨을 debug로 세팅합니다.")  # "store_true" means "set default to False"
+        buy_lotto645.add_argument("-d", "--debug", action="store_true",
+                                  help="로그 출력 레벨을 debug로 세팅합니다.")  # "store_true" means "set default to False"
         self._args = parser.parse_args()
 
-        if not self._args.username is None:
+        if self._args.username:
```
Comment (roeniss) [inline src/dhapi/router/arg_parser.py:83]: cool

File: src/dhapi/router/arg_parser.py [hunk 3]
```diff
@@ -51,7 +51,8 @@ def __init__(self):
         buy_lotto645.formatter_class = argparse.RawTextHelpFormatter
 
         buy_lotto645.add_argument("-u", "--username", required=False, help="동행복권 아이디입니다. (deprecated; -p 옵션 사용 권장)")
-        buy_lotto645.add_argument("-q", "--quiet", action="store_true", help="플래그 설정 시 구매 전 확인 절차를 스킵합니다.")  # "store_true" means "set default to False"
+        buy_lotto645.add_argument("-q", "--quiet", action="store_true",
+                                  help="플래그 설정 시 구매 전 확인 절차를 스킵합니다.")  # "store_true" means "set default to False"
```
Comment (roeniss) [inline src/dhapi/router/arg_parser.py:55]: This line should be rollbacked. I see no value here

**Review by roeniss — CHANGES_REQUESTED** [review]
The basic idea looks fine. Thank you for this pr.

But I don't want to use your own package. There are several reasons:

1. Your package doesn't use versioning. This could bring unexpected error to dhapi .
2. "core_util" means nothing. What 'core' do you mean? How "config file read program"  become "core"? 
3. This doesn't follow dhapi's error handling flow. 

nevertheless, the code seems having good abstraction. Why don't you implement it dhapi files.

**Review by roeniss — CHANGES_REQUESTED** [review]
The basic idea looks fine. Thank you for this pr.

But I don't want to use your own package. There are several reasons:

1. Your package doesn't use versioning. This could bring unexpected error to dhapi .
2. "core_util" means nothing. What 'core' do you mean? How "config file read program"  become "core"? 
3. This doesn't follow dhapi's error handling flow. 

nevertheless, the code seems having good abstraction. Why don't you implement it dhapi files.

File: README.md (hunk 1 above)
Comment (seunggabi) [reply README.md:3]: Jetbrains tool(pycharm) suggest `shell` (not `sh`)
스크린샷 2023-06-06 오후 4 03 54

So, I think of this expression common.

File: src/dhapi/router/arg_parser.py (hunk 3 above)
Comment (seunggabi) [reply src/dhapi/router/arg_parser.py]: got it.
just IDE auto adding new line.. haha

Comment (seunggabi) [reply src/dhapi/router/arg_parser.py]: 스크린샷 2023-06-06 오후 4 10 33
스크린샷 2023-06-06 오후 4 12 48

Do you apply `max-line-length` (default: 120) ?

Comment (roeniss) [reply src/dhapi/router/arg_parser.py]: It's 180, which is configured by [.pylintrc](https://github.com/roeniss/dhlottery-api/blob/main/.pylintrc).

Please disable your IDE formatting and run `make lintfmt` after last commit. See this: https://github.com/roeniss/dhlottery-api/blob/main/docs/CONTRIBUTING.md#pr-%EC%A0%84-%ED%99%95%EC%9D%B8%EC%82%AC%ED%95%AD

File: README.md (hunk 1 above)
Comment (roeniss) [reply README.md:3]: If that's the case, I would rather suggest `sh`. I didn't see any `shell` in my life (until now). Can you rollback this?

**Review by roeniss — CHANGES_REQUESTED** [review]
Please check my comment, and run `make lintfmt`. After that this pr will be merged (please request review after doing that)

File: README.md (hunk 1 above)
Comment (seunggabi) [reply README.md:3]: https://docs.readme.com/rdmd/docs/code-blocks
스크린샷 2023-06-07 오후 1 23 37

`bash` 오픈 소스들 보면 이게 많이 쓰이고
`sh` 로 돌려둘게요~

File: src/dhapi/router/arg_parser.py (hunk 3 above)
Comment (seunggabi) [reply src/dhapi/router/arg_parser.py]: editor 에서 pylintrc 를 인식하지? 않네요..ㅠ 감사합니다.

File: README.md (hunk 1 above)
Comment (seunggabi) [reply README.md:3]: 이견 없습니다. shoooort is goooood

**Review by roeniss — APPROVED** [review]
@seunggabi LGTM.

But I didn't test this code yet (use all quota of this week) neither does this pr include any test. So, I will verify it by myself next week, then if there's no problem, this pr will be merged and a newer version will be published. 

Again, thank you for your contribution. Your job means a lot to me.

**Comment by seunggabi** [comment]
@roeniss thx verrrry much.

**Comment by roeniss** [comment]
@seunggabi I've found a few things to change.

1. Please check `make lintfmt` to get 10/10 points at the end. when I run it, I get 9.96/10

```
$ make lintfmt
black src/**
All done! ✨ 🍰 ✨
16 files left unchanged.
pylint src
************* Module dhapi.router.arg_parser
src/dhapi/router/arg_parser.py:8:0: C0411: third party import "from seunggabi_core_python.util import config_util" should be placed before "from dhapi.router.version_checker import get_versions" (wrong-import-order)

------------------------------------------------------------------
Your code has been rated at 9.96/10 (previous run: 9.96/10, +0.00)

make: *** [lintfmt] Error 16

```

2. Return proper error message when the credential file have a problems:
  - this is basically the same comment with [this](https://github.com/roeniss/dhlottery-api/pull/45#pullrequestreview-1461076151) (the third point)
  - the file not exists (~/.dhapi/credentials
  - the format is wrong (e.g., omit `username = ` and `password = `)
  - in any cases, the content in credential file must not be revealed in stdout (terminal)

Below logs shows the example:

```
// ~/.dhapi/credentails
[default]
hello
world

// the execution result
$ PYTHONPATH=./src/ python3 src/dhapi/main.py buy_lotto645
Traceback (most recent call last):
  File "/Users/roeniss/tmp/pr45/src/dhapi/main.py", line 9, in <module>
    entrypoint()
  File "/Users/roeniss/tmp/pr45/src/dhapi/router/router.py", line 7, in entrypoint
    arg_parser = ArgParser()
                 ^^^^^^^^^^^
  File "/Users/roeniss/tmp/pr45/src/dhapi/router/arg_parser.py", line 86, in __init__
    credentials = config_util.get(group="dhapi", context="credentials", profile=self._args.profile[0])
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/roeniss/tmp/pr45/docs/.venv/lib/python3.11/site-packages/seunggabi_core_python/util/config_util.py", line 30, in get
    config.read(f"{home()}/{path(group, context)}")
  File "/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/configparser.py", line 713, in read
    self._read(fp, filename)
  File "/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/configparser.py", line 1132, in _read
    raise e
configparser.ParsingError: Source contains parsing errors: '/Users/roeniss/.dhapi/credentials'
	[line  2]: 'hello\n'
	[line  3]: 'world\n'
```

All of the above logs should be replaced with something like this: 

```
🚨 에러가 발생했습니다: 로그인에 실패했습니다. ~/.dhapi/credentials 파일을 확인해주세요.
```

Use [_exit](https://github.com/roeniss/dhlottery-api/blob/5e44ca7620eb92275a82af97ad60bae1eee8dbbf/src/dhapi/router/arg_parser.py#L17-L19) functions after call `config_util.get()`. As I mentioned above, you should distinguish each error case to inform the exact problem to the user.

**Comment by roeniss** [comment]
If you're not sure what to do, feel free to mention me. I'll handle it after merge!

**Comment by seunggabi** [comment]
@roeniss 
### fixed
- lintfmt
```sh
(venv)  seunggabi@gimseung-gab-ui-MacBookPro  ~/seunggabi/pr/pr-dhlottery-api   feature/#44_credentials  make lintfmt
black src/**
All done! ✨ 🍰 ✨
16 files left unchanged.
pylint src

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```
- not exposed
```sh
(venv)  seunggabi@gimseung-gab-ui-MacBookPro  ~/seunggabi/pr/pr-dhlottery-api   feature/#44_credentials  PYTHONPATH=./src/ python3 src/dhapi/main.py buy_lotto645 -p asdf
Traceback (most recent call last):
  File "/Users/seunggabi/seunggabi/pr/pr-dhlottery-api/src/dhapi/main.py", line 9, in <module>
    entrypoint()
  File "/Users/seunggabi/seunggabi/pr/pr-dhlottery-api/src/dhapi/router/router.py", line 7, in entrypoint
    arg_parser = ArgParser()
  File "/Users/seunggabi/seunggabi/pr/pr-dhlottery-api/src/dhapi/router/arg_parser.py", line 84, in __init__
    credentials = config_util.get(group="dhapi", context="credentials", profile=self._args.profile[0])
  File "/Users/seunggabi/seunggabi/pr/pr-dhlottery-api/venv/lib/python3.9/site-packages/seunggabi_core_python/util/config_util.py", line 35, in get
    raise NotFound(p) from None
seunggabi_core_python.exception.not_found.NotFound: [404 NOT_FOUND] /Users/seunggabi/.dhapi/credentials

```

**Review by roeniss — CHANGES_REQUESTED** [review]
Actually, your new error-handling is not what I intended. 

See [here](https://github.com/roeniss/dhlottery-api/blob/main/src/dhapi/router/arg_parser.py#L13) and [here](https://github.com/roeniss/dhlottery-api/blob/main/src/dhapi/configuration/logger.py#L25).

Your error should say like this: 

```sh
❯❯❯ dhapi buy_23
🚨 입력 파라미터 에러 발생: argument command: invalid choice: 'buy_23' (choose from 'buy_lotto645')

❯❯❯ dhapi buy_lotto645
2023-06-12 15:39:30 [error    ] 🚨 에러가 발생했습니다: 동행복권 사이트가 현재 시스템 점검중입니다.
🚨 에러가 발생했습니다: 동행복권 사이트가 현재 시스템 점검중입니다.
```

the second is because it's 1 a.m. now... anyway, I think you need to implement it like this: https://github.com/roeniss/dhlottery-api/blob/main/src/dhapi/client/lottery_client.py#L48

You have two choices:

1. wrap each error with `raise RuntimeError()` in your own library: In this case, you retain the current pull request code and update your own library only.
2. wrap `credentials = config_util.get(...)` line with `raise RuntimeError()`: In this case, you would need to add some conditional statements to distinguish each error, but instead your library would be independent to my program.

**Comment by seunggabi** [comment]
@roeniss 
이해가 안되는데 혹시 코드로 부탁드려도 될까요?
흠.. profile 과 에러는 무관한듯 보이는데.. ㅎㅎ
그건 별도 이슈로 처리하는게 어떨까합니다.

**Comment by roeniss** [comment]
@seunggabi 

> 그건 별도 이슈로 처리하는게 어떨까합니다. 

네 좋습니다. 최근 이 프로젝트에 신경을 못쓰고 있는데요, 최대한 시간을 내어 테스트 후 머지를 진행하겠습니다. 아래 이슈 때문에 정상적으로 테스트를 못하고 있습니다 :bow:


- https://github.com/roeniss/dhlottery-api/issues/47

**Comment by roeniss** [comment]
- resolves https://github.com/roeniss/dhlottery-api/pull/45

**Pull request merged**